src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs

```diff
@@ -1777,6 +1779,60 @@ internal void EnsureNativeIntegerAttributeExists()
             EnsureEmbeddableAttributeExists(EmbeddableAttributes.NativeIntegerAttribute);
         }
 
+#nullable enable
+        /// <summary>
+        /// Creates the ThrowIfNull and Throw helpers if needed.
+        /// </summary>
+        /// <remarks>
+        /// The ThrowIfNull and Throw helpers are modeled off of the helpers on ArgumentNullException.
+        /// https://github.com/dotnet/runtime/blob/22663769611ba89cd92d14cfcb76e287f8af2335/src/libraries/System.Private.CoreLib/src/System/ArgumentNullException.cs#L56-L69
+        /// </remarks>
+        internal MethodSymbol EnsureThrowIfNullFunctionExists(SyntaxNode syntaxNode, SyntheticBoundNodeFactory factory, DiagnosticBag? diagnostics)
+        {
+            // If we have already generated the helper, possibly for another switch
+            // or on another thread, we don't need to regenerate it.
+            var privateImplClass = GetPrivateImplClass(syntaxNode, diagnostics);
+            if (getExistingThrowIfNullMethod() is { } existingMethod)
+            {
+                return existingMethod;
+            }
+
+            TypeSymbol returnType = factory.SpecialType(SpecialType.System_Void);
+            TypeSymbol argumentType = factory.SpecialType(SpecialType.System_Object);
+            TypeSymbol paramNameType = factory.SpecialType(SpecialType.System_String);
+
+            var sourceModule = SourceModule;
+            var throwMethod = new SynthesizedThrowMethod(sourceModule, privateImplClass, returnType, paramNameType);
+            privateImplClass.TryAddSynthesizedMethod(throwMethod.GetCciAdapter());
+
+            // If this call lost a race with another call, we want to obtain and return the symbol that was actually added successfully.
+            // note: strictly it is possible for another thread to add the 'throwMethod', then suspend, then this thread adds the 'throwIfNullMethod'.
+            // If this situation occurs it is not really a concern. We will emit all the methods we need and even if we use a different object instance
```

Comment:
>If this situation occurs it is not really a concern. [](http://example.com/codeflow?start=15&length=52)

> If this situation occurs it is not really a concern.

I don't agree. If we ever store these things in a dictionary, or ever expose them through a public API, we have a nasty bug waiting for us. Let's make sure that we're not accidentally introducing bad race conditions here.